src/modules/fancyzones/lib/FancyZonesData.cpp

```diff
@@ -0,0 +1,609 @@
+        {
+            MigrateCustomZoneSetsFromRegistry();
+
+            SaveFancyZonesData();
+        }
+        else
+        {
+            json::JsonObject fancyZonesDataJSON = GetPersistFancyZonesJSON();
+
+            appZoneHistoryMap = JSONHelpers::ParseAppZoneHistory(fancyZonesDataJSON);
+            deviceInfoMap = JSONHelpers::ParseDeviceInfos(fancyZonesDataJSON);
+            customZoneSetsMap = JSONHelpers::ParseCustomZoneSets(fancyZonesDataJSON);
+        }
+    }
+
+    void FancyZonesData::SaveFancyZonesData() const
+    {
+        std::scoped_lock lock{ dataLock };
+        json::JsonObject root{};
+        json::JsonObject appZoneHistoryRoot{};
+
+        appZoneHistoryRoot.SetNamedValue(L"app-zone-history", JSONHelpers::SerializeAppZoneHistory(appZoneHistoryMap));
+        root.SetNamedValue(L"devices", JSONHelpers::SerializeDeviceInfos(deviceInfoMap));
+        root.SetNamedValue(L"custom-zone-sets", JSONHelpers::SerializeCustomZoneSets(customZoneSetsMap));
+
+        auto before = json::from_file(zonesSettingsFileName);
+        if (!before.has_value() || before.value().Stringify() != root.Stringify())
+        {
+            Trace::FancyZones::DataChanged();
+        }
+
+        json::to_file(zonesSettingsFileName, root);
+        json::to_file(appZoneHistoryFileName, appZoneHistoryRoot);
+    }
+
+    void FancyZonesData::MigrateCustomZoneSetsFromRegistry()
+    {
+        std::scoped_lock lock{ dataLock };
+        wchar_t key[256];
+        StringCchPrintf(key, ARRAYSIZE(key), L"%s\\%s", REG_SETTINGS, L"Layouts");
+        HKEY hkey;
+        if (RegOpenKeyExW(HKEY_CURRENT_USER, key, 0, KEY_ALL_ACCESS, &hkey) == ERROR_SUCCESS)
+        {
+            BYTE data[256];
+            DWORD dataSize = ARRAYSIZE(data);
+            wchar_t value[256]{};
+            DWORD valueLength = ARRAYSIZE(value);
+            DWORD i = 0;
+            while (RegEnumValueW(hkey, i++, value, &valueLength, nullptr, nullptr, reinterpret_cast<BYTE*>(&data), &dataSize) == ERROR_SUCCESS)
+            {
+                FancyZonesDataTypes::CustomZoneSetData zoneSetData;
+                zoneSetData.name = std::wstring{ value };
+                zoneSetData.type = static_cast<FancyZonesDataTypes::CustomLayoutType>(data[2]);
+                // int version =  data[0] * 256 + data[1]; - Not used anymore
+
+                GUID guid;
+                auto result = CoCreateGuid(&guid);
+                if (result != S_OK)
+                {
+                    continue;
+                }
+                wil::unique_cotaskmem_string guidString;
+                if (!SUCCEEDED_LOG(StringFromCLSID(guid, &guidString)))
+                {
+                    continue;
+                }
+
+                std::wstring uuid = guidString.get();
+
+                switch (zoneSetData.type)
+                {
+                case FancyZonesDataTypes::CustomLayoutType::Grid:
+                {
+                    int j = 5;
```

Comment:
Can you add a comment to explain why `j = 5`?

Reply:
I could only put these links in the comments (FZ v.14 PersistData() functions):
https://github.com/microsoft/PowerToys/blob/v0.14.0/src/modules/fancyzones/editor/FancyZonesEditor/Models/GridLayoutModel.cs#L147
https://github.com/microsoft/PowerToys/blob/v0.14.0/src/modules/fancyzones/editor/FancyZonesEditor/Models/CanvasLayoutModel.cs#L126

if that's ok?

Reply:
Not sure I understand how those links help to understand why j = 5.

Reply:
That's how custom layouts data was packed in registry back then, so that's how we read it from registry when we do the migration

Reply:
OK, given we are not going to change that code, I guess the links should be enough.

Reply:
Comments added. Please, check if ok

Reply:
OK